### PR TITLE
resolve NPE in NetHandlerPlayClient#handleSpawnPlayer for both 1.8.9 …

### PR DESCRIPTION
**Pull request by Alexdoru**
…& 1.12.2

image

### CODE REVIEW

**Review by Sychic — COMMENTED** [review]
Is this impossible to do with a mixin?

**Comment by Alexdoru** [comment]
> Is this impossible to do with a mixin?

One could do an inject before, check the hashmap before hand and if it returns null do callback.cancel() but I didn't feel like checking the hasmap twice. Other than that I don't see how this could be done with mixins.

**Comment by DJtheRedstoner** [comment]
Mixin should be used for all new injections, unless it's absolutely necessary to use ASM.

**Comment by Alexdoru** [comment]
> Mixin should be used for all new injections, unless it's absolutely necessary to use ASM.

is it possible to fix the bug without checking the haskmap twice ? like is there a mixin injector that allows you to catch the return value of a certain method invokation and then decide to cancel the whole method (inject a return) based on that

**Comment by DJtheRedstoner** [comment]
No, just check the hashmap twice. Lookups are O(1).

**Comment by Alexdoru** [comment]
> No, just check the hashmap twice. Lookups are O(1).

it's done

File: src/main/java/club/sk1er/patcher/mixins/bugfixes/NetHandlerPlayClient_FixHandleSpawnPlayerNPE.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
+package club.sk1er.patcher.mixins.bugfixes;
+
+import net.minecraft.client.network.NetHandlerPlayClient;
+import net.minecraft.client.network.NetworkPlayerInfo;
+//#if MC==10809
+import net.minecraft.network.play.server.S0CPacketSpawnPlayer;
```
Comment (DJtheRedstoner) [inline src/main/java/club/sk1er/patcher/mixins/bugfixes/NetHandlerPlayClient_FixHandleSpawnPlayerNPE.java:6]: 1. Imports with preprocessor statements should be placed below regular imports.
2. If you add `net.minecraft.network.play.server.SPacketSpawnPlayer net.minecraft.network.play.server.S0CPacketSpawnPlayer` to `versions/1.12.2-1.8.9.txt`, preprocessor isn't needed in this file. Put this line with the rest of the packet mappings.

Comment (Alexdoru) [reply src/main/java/club/sk1er/patcher/mixins/bugfixes/NetHandlerPlayClient_FixHandleSpawnPlayerNPE.java:5]: it's done

**Review by Sychic — APPROVED** [review]
LGTM

**Review by DJtheRedstoner — APPROVED** [review]

**Pull request merged**